cumulus/client/consensus/common/Cargo.toml

```
@@ -10,7 +10,7 @@ async-trait = "0.1.73"
 codec = { package = "parity-scale-codec", version = "3.0.0", features = [ "derive" ] }
 dyn-clone = "1.0.12"
 futures = "0.3.28"
-log = "0.4.20"
+log = { workspace = true, default-features = true }
```

```suggestion
log = { workspace = true }
```

This and all the other places you have done this "feel weird"

---

Yea I think the `default-features = true` should be implicit in this case, so we should be able to omit it.

---

:+1: 

I think it makes sense to do `default-features = false` in the workspace `Cargo.toml` *on a case-by-case basis* if we actually don't want to always enable all of the features for a given dependency, but this probably shouldn't be done by default (at least not blindly), and if you're adding a `default-features = true` all over the place then it's most likely the wrong thing to do.

---

And in case of the `log` [it looks like it doesn't even have any default features](https://github.com/rust-lang/log/blob/master/Cargo.toml).

---

>And in case of the log [it looks like it doesn't even have any default features](https://github.com/rust-lang/log/blob/master/Cargo.toml).

Hm. Then we would need to check on every version update whether some default features were added? I guess that could work since the WASM build should fail when it tries to pull in std.

>and if you're adding a `default-features = true` all over the place then it's most likely the wrong thing to do. 

Yes I agree if this is the case in 100% of the occurences (and under the assumption that this would not change in the future). But for a dependency where we sometimes want default-features and sometimes not, then we need to set it to false in the workspace i think.  

> If a crate needs a specific feature it should explicitly say which one, instead of doing ` default-features = true`  

If this is the goal, then i dont think that it is automatable.

---

> But for a dependency where we sometimes want default-features and sometimes not, then we need to set it to false in the workspace i think.

I don't think so. Cargo unifies features and the same will probably happen here. But for things crates that we use in no_std and somewhere only in std, we should Kot disable the default features in the workspace. IMO.

We should read up on how feature setting works with workspace deps. Especially around default features.

---

I disabled the default features in the workspace after reading this: https://github.com/rust-lang/cargo/issues/11329  
And the unification does not apply when crates are build on their own, or?  
I can try to find some more resources around the topic, but will probably close the MR then for now.

---

What a fun :heart:

---

I ran into a similar thing when I tried this; you need to specify `default-features = false` in the worksapce in order to be able to use `default-features` on that dependency in the crate's Cargo.toml, else you get a warning like:

```
warning: ..polkadot-sdk/polkadot/xcm/Cargo.toml: `default-features` is ignored for scale-info, since `default-features` was not specified for `workspace.dependencies.scale-info`, this could become a hard error in the future
```

---

Yes, this is what @ggwpez linked above.

---

So can we go ahead with this:  
- `default-features = false` in the root workspace
- nothing in the crates if the default-features should be true
- `default-features = false` in each crate like before when they should be disabled.

This should make the diff here a bit shorter.

---

If that works, lets do this.